feat(portable): .NET46 Storage Module

GantMan opened the pull request:
### Storage Module
* Success migrating .NET46 storage to use [PCLStorage](http://www.nudoq.org/#!/Packages/PCLStorage).   
* Windows Universal is still using [Windows.Storage](https://msdn.microsoft.com/en-us/library/windows/apps/windows.storage.aspx).
* Helper class `AsyncStorageErrorHelper.cs` is migrated to shared.

cc: @matthargett - @kevinvangelder

matthargett commented:
If @rozele is cool with a TODO comment, then LGTM

rozele commented on ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj:
```diff
@@ -66,6 +74,7 @@
     <Reference Include="System.Xml" />
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="Modules\Storage\AynceStorageModule.cs" />
```
>AynceStorageModule [](start = 38, length = 18)

Weird filename, Aynce -> Async

GantMan commented on ReactWindows/ReactNative.Net46/Modules/Storage/AynceStorageModule.cs:
```diff
@@ -0,0 +1,493 @@
+
+namespace ReactNative.Modules.Storage
+{
+    class AsyncStorageModule : NativeModuleBase, ILifecycleEventListener
```
This could be a perfect place for that.  I'll give that a go, and let's see what it looks like.  If it ends up being distracting, or difficult to the developer, I'll let you know.

GantMan commented:
@rozele - as you can see, I made a new branch and moved everything into shared with an adapter.

https://github.com/infinitered/react-native-windows/compare/shared-project-storage-module...infinitered:shared-project-storage-module-adapter?expand=1

Is this the abstraction you're looking for?  Shared code needs preprocessor directives though.
@matthargett - thoughts?

GantMan commented:
@rozele - moved some constants to shared.  I think this is a clean division between the two.  Please review for merge

rozele commented:
I'd recommend keeping the constants as private fields and moving the helper methods for normalizing key and file names as well.

---
In reply to: [252665901](https://github.com/ReactWindows/react-native-windows/pull/748#issuecomment-252665901) [](ancestors = 252665901)

GantMan commented:
@rozele - done

rozele commented on ReactWindows/ReactNative.Shared/Modules/Storage/AsyncStorageConstants.cs:
```diff
@@ -0,0 +1,135 @@
+        public const string DirectoryName = "AsyncStorage";
+        public const string FileExtension = ".data";
+
+        public static readonly IDictionary<char, string> s_charToString = new Dictionary<char, string>
```
private static

rozele commented on ReactWindows/ReactNative.Shared/Modules/Storage/AsyncStorageConstants.cs:
```diff
@@ -0,0 +1,135 @@
+            { '}', "{ccb}" },
+        };
+
+        public static readonly IDictionary<string, char> s_stringToChar = s_charToString.ToDictionary(kv => kv.Value, kv => kv.Key);
```
>public [](start = 8, length = 6)

private static

rozele commented on ReactWindows/ReactNative.Shared/Modules/Storage/AsyncStorageConstants.cs:
```diff
@@ -0,0 +1,135 @@
+
+        public static readonly IDictionary<string, char> s_stringToChar = s_charToString.ToDictionary(kv => kv.Value, kv => kv.Key);
+
+        public static readonly int s_maxReplace = s_charToString.Values.Select(s => s.Length).Max();
```
>public [](start = 8, length = 6)

private static

rozele commented on ReactWindows/ReactNative.Shared/Modules/Storage/AsyncStorageConstants.cs:
```diff
@@ -0,0 +1,135 @@
+
+namespace ReactNative.Modules.Storage
+{
+    class AsyncStorageConstants
```
>AsyncStorageConstants [](start = 10, length = 21)

Constants is not really a good name for this anymore... maybe just consolidate it with AsyncStorageErrorHelpers and call it AsyncStorageHelpers?

GantMan commented:
### Update for @rozele 
* Moved *ALL* helpers to new shared `AsyncStorageHelpers.cs` (which looks/feels much cleaner)
* Cleaned up noted scope changes
* merged with master

rozele commented on ReactWindows/ReactNative.Shared/Modules/Storage/AsyncStorageHelpers.cs:
```diff
@@ -0,0 +1,158 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json.Linq;
```
Sort namespaces.

rozele commented:
Looks fantastic! Just one minor comment about sorting the namespaces in AsyncStorageHelpers.cs and this is good to go!

---
In reply to: [252970683](https://github.com/ReactWindows/react-native-windows/pull/748#issuecomment-252970683) [](ancestors = 252970683)

rozele commented:
:shipit:

rozele commented:
Sorry @GantMan this one needs to be rebased

GantMan commented:
@rozele 👍  should be good now.
